rust/src/countme.rs

```diff
@@ -0,0 +1,139 @@
+/*
+ * Copyright (C) 2020 Red Hat, Inc.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA
+ */
+
+use anyhow::{bail, Result};
+use curl::easy::Easy;
+use os_release::OsRelease;
+use std::path;
+
+mod cookie;
+mod repo;
+
+// Default variant name used in User Agent
+const DEFAULT_VARIANT_ID: &str = "unknown";
+
+// Send a GET request to 'url' with 'ua' as User Agent and discard the response
+fn send_countme(url: &str, ua: &str) -> Result<()> {
+    println!("Sending request to: {}", url);
+    let mut handle = Easy::new();
+    handle.url(&url)?;
+    handle.useragent(&ua)?;
+    {
+        let mut transfer = handle.transfer();
+        transfer.write_function(|new_data| Ok(new_data.len()))?;
+        transfer.perform()?;
+    }
+    Ok(())
+}
+
+/// Main entrypoint for countme
+fn countme_main(_args: &[String]) -> Result<()> {
+    // Silently skip if we are not run on an ostree booted system
+    if !path::Path::new("/run/ostree-booted").exists() {
+        println!("Skipping: Not running on an ostree based system.");
+        return Ok(());
+    }
+
+    // Load repo configs and keep only those enabled, with a metalink and countme=1
+    let repos: Vec<_> = self::repo::all()?
+        .into_iter()
+        .filter(|r| r.count_me())
+        .collect();
+    if repos.is_empty() {
+        println!("No enabled repositories with countme=1");
+        return Ok(());
+    }
+
+    // Load timestamp cookie
+    let cookie = match cookie::Cookie::new() {
+        Err(e) => bail!("Could not read existing cookie: {}", e),
+        Ok(c) => c,
+    };
```

Comment:
Minor: I think the more conventional idiom is

```suggestion
    let cookie = cookie::Cookie::new().context("could not read existing cookie")?;
```

?